### PR TITLE
unicode-range may cause us to use Times's font metrics erroneously

### PR DESCRIPTION
**Pull request by litherum**
#### 7ba7f4d9fb851ed545353bb489d4b929f7f5e8cd
<pre>
unicode-range may cause us to use Times&apos;s font metrics erroneously
<a href="https://bugs.webkit.org/show_bug.cgi?id=244757">https://bugs.webkit.org/show_bug.cgi?id=244757</a>

Reviewed by Darin Adler.

When we lay out a line, we use the metrics of the primary font inside our computation to determine line height.
If the content just says `font-family: a, b`; then it&apos;s clear that `a` is the primary font. However, `a` can
actually be a composite font, as such:

```
@font-face {
    font-family: a;
    src: url(&quot;MyFont.ttf&quot;);
    unicode-range: some range;
}
@font-face {
    font-family: a;
    src: url(&quot;MyOtherFont.ttf&quot;);
    unicode-range: a distinct range that doesn&apos;t intersect the other font&apos;s range;
}
```

In this situation, when laying out a line, should we use the metrics from MyFont.ttf or MyOtherFont.ttf?

The spec isn&apos;t prescriptive here, so it&apos;s kind of up to us to determine what to do. What we actually do is
from FontCascadeFonts::primaryFont():
A) If one of the `unicode-range`s includes the space character, use that one (last one wins if there are
      multiple). If we need to download this font, do so.
B) If none of the `unicode-range`s includes the space character, then use the last @font-face block, but
      set the flag to forbid downloads from occuring.

When we ask for a font that hasn&apos;t finished loading yet, the font object we get back is an &quot;interstitial&quot;
font, which just means it&apos;s a temporary invisible font to use until the real one finishes downloading. The
interstitial font is hardcoded to always be Times. When we set the flag to forbid downloads from occurring,
the font we get back is this &quot;interstitial&quot; font, but we&apos;ll never switch from the interstitial font to the
real one - because the real one will never finish downloading because it never started downloading! So we
just use the metrics from Times indefinitely, which is a surprising result!

The solution here is just to allow downloads in both cases A) and B) above. By allowing downloads in case
B), we start the font loading process which means we&apos;ll eventually switch off of using Times&apos;s metrics.

Test: http/tests/text/primary-font-interstitial.html

* LayoutTests/http/tests/text/primary-font-interstitial-expected.html: Added.
* LayoutTests/http/tests/text/primary-font-interstitial.html: Added.
* LayoutTests/http/tests/text/resources/Gulf-regular.ttf: Added.
* Source/WebCore/platform/graphics/FontCascadeFonts.h:
(WebCore::FontCascadeFonts::primaryFont):

Canonical link: <a href="https://commits.webkit.org/262464@main">https://commits.webkit.org/262464@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/61950fdef699f0d6f722632bf8b9ec0e7249d1b0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1588 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1619 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1676 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2509 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/1712 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1684 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1680 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1492 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1602 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1442 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1437 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/2352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1441 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1423 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1341 "52 flakes 182 failures") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1446 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/1463 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/2505 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1525 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1482 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/1629 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1398 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1430 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/351 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/394 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1545 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/1674 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/466 "Passed tests") |

### CODE REVIEW

File: Source/WebCore/platform/graphics/FontRanges.h [hunk 1]
```diff
@@ -61,7 +61,7 @@ class FontRanges {
 
         UChar32 from() const { return m_from; }
         UChar32 to() const { return m_to; }
-        const Font* font(ExternalResourceDownloadPolicy) const;
+        WEBCORE_EXPORT const Font* font(ExternalResourceDownloadPolicy) const;
```
Comment (darinadler) [inline Source/WebCore/platform/graphics/FontRanges.h:64]: Why this change? Could you discard this?

**Review by darinadler — APPROVED** [review]

**Comment by litherum** [comment]
> @litherum - This one never landed. Just wanted to check old PRs, where we are passing tests etc. but didn't landed. Sorry if I nudged your unnecessarily.

Yep, I still need to land it.

File: Source/WebCore/platform/graphics/FontRanges.h (hunk 1 above)
Comment (litherum) [reply Source/WebCore/platform/graphics/FontRanges.h:64]: This change is because the above change in `FontCascadeFonts.h` is inlined, which means other frameworks will now try to call `FontRanges::font()`.
